File: IT/2/Multithreading/Java/Multithreading3.java
/* file Multithreading3.java */

import java.util.concurrent.ThreadLocalRandom;

/* le classi MainFunc e SelectFromDb implementano l'interfaccia Runnable, e definiscono il metodo run() */

class SelectFromDb implements Runnable {
	@Override
	public void run() {
		Thread t = Thread.currentThread();
		String tName = t.getName();
		System.out.println(tName + " - selezione database");

		try {
			for (int i = 0; i < 3; i++) {
				System.out.println(tName + " - elaborazione dati...");
				// pausa da 1 a 3 secondi
				int pause = ThreadLocalRandom.current().nextInt(1, 3+1);
				t.sleep(pause * 1000);
			}
		} catch (InterruptedException e) {
			System.err.println(e);
		}
	}
}

class MainFunc implements Runnable {
	@Override
	public void run() {
		Thread t = Thread.currentThread();
		String tName = t.getName();

		try {
			for (int i = 0; i < 3; i++) {
				System.out.println(tName + " - ciclo principale");
				// pausa da 1 a 3 secondi
				int pause = ThreadLocalRandom.current().nextInt(1, 3+1);
				t.sleep(pause * 1000);
			}
		} catch (InterruptedException e) {
			System.err.println(e);
		}
	}
}

public class Multithreading3 {
	public static void main(String [] args) {
		try {
			/* crea due nuovi thread di tipo Runnable */

			MainFunc process = new MainFunc();
			Thread t1 = new Thread(process, "MAIN");

			SelectFromDb select = new SelectFromDb();
			Thread t2 = new Thread(select, "BACKGROUND");

			/* abilita i nuovi thread */

			t1.start();
			t2.start();

			t1.join();
			t2.join();

			System.out.println("Dati dal database pronti all'uso!");
		}
		catch (InterruptedException e) {
			System.err.println(e);
		}
	}
}

File: IT/2/Multithreading/C++/multithreading3.cpp
/*
file multithreading3.cpp
compilare con: g++ -std=c++11 multithreading3.cpp -lpthread
*/

#include <cstdlib>
#include <random>
#include <chrono>
#include <thread>

using std::random_device;
using std::default_random_engine;
using std::uniform_int_distribution;
using std::chrono::seconds;
using std::this_thread::sleep_for;
using std::thread;

void selectFromDb(const char* tname)
{
	printf("%s - selezione database\n", tname);

	thread_local default_random_engine gen(random_device{}());

	for (int i = 0; i < 3; i++) {
		printf("%s - elaborazione dati...\n", tname);
		// pausa da 1 a 3 secondi
		uniform_int_distribution<int> dist(1, 3);
		seconds duration(dist(gen));
		sleep_for(duration);
	}
}

/* l'invocazione a this_thread::sleep_for() lascia il processore al thread che esegue la select */

void mainFunc(const char* tname)
{
	thread_local default_random_engine gen(random_device{}());

	for (int i = 0; i < 3; i++) {
		printf("%s - ciclo principale\n", tname);
		// pausa da 1 a 3 secondi
		uniform_int_distribution<int> dist(1, 3);
		seconds duration(dist(gen));
		sleep_for(duration);
	}
}

int main() 
{
	/* crea due nuovi thread */

	thread t1(mainFunc, "MAIN");
	thread t2(selectFromDb, "BACKGROUND");

	t1.join();
	t2.join();

	printf("Dati dal database pronti all'uso!\n");
}

